src/utils.rs

```diff
@@ -1,28 +1,74 @@
 use crate::Error;
+use libflate::gzip;
 use std::collections::HashMap;
+use std::io::{self, Read};
 use walrus::*;
 
+pub const WASM_MAGIC_BYTES: &[u8] = &[0, 97, 115, 109];
+
+pub const GZIPPED_WASM_MAGIC_BYTES: &[u8] = &[31, 139, 8, 0];
+
 fn wasm_parser_config(keep_name_section: bool) -> ModuleConfig {
     let mut config = walrus::ModuleConfig::new();
     config.generate_name_section(keep_name_section);
     config.generate_producers_section(false);
     config
 }
 
-pub fn parse_wasm(wasm: &[u8], keep_name_section: bool) -> Result<Module, Error> {
+pub fn parse_wasm(bytes: &[u8], keep_name_section: bool) -> Result<Module, Error> {
+    let wasm = if bytes.starts_with(WASM_MAGIC_BYTES) {
+        Ok(bytes)
+    } else if bytes.starts_with(GZIPPED_WASM_MAGIC_BYTES) {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Please use parse_wasm_robust with gzipped inputs.",
```

Comment:
This doesn't seem to be a breaking change, why not just decompress here, instead of a new function?

Reply:
The original public function `parse_wasm`'s signature has `bytes: &[u8]`, changing it to `bytes: Vec<u8>` would be a breaking change, right?

If we keep it `&[u8]`, then the `decompress` function's result type (`Result<Vec<u8>`) is incompatible with it in one of the branches. So either we would need to inline `decompress` or have another function that takes `Vec<u8>` to begin with.

What's your preference? Am I missing a better possible implementation?

Reply:
> If we keep it &[u8], then the decompress function's result type (Result<Vec<u8>) is incompatible with it in one of the branches.

We can make both branches to return `Cow::Borrowed` and `Cow::Owned`, so that they end up with the same type.

Reply:
Let me give it a try

Reply:
Seems to have worked right away. Thanks for the tip!